client/allocrunner/tasklifecycle/coordinator.go

```diff
@@ -0,0 +1,415 @@
+		return "wait_main"
+	case coordinatorStatePoststop:
+		return "poststart"
+	}
+	return ""
+}
+
+// lifecycleStage represents a lifecycle configuration used for task
+// coordination.
+//
+// Not all possible combinations of hook X sidecar are defined, only the ones
+// that are relevant for coordinating task initialization order. For example, a
+// main task with sidecar set to `true` starts at the same time as a
+// non-sidecar main task, so there is no need to treat them differently.
+type lifecycleStage uint8
+
+const (
+	// lifecycleStagePrestartEphemeral are tasks with the "prestart" hook and
+	// sidecar set to "false".
+	lifecycleStagePrestartEphemeral lifecycleStage = iota
+
+	// lifecycleStagePrestartSidecar are tasks with the "prestart" hook and
+	// sidecar set to "true".
+	lifecycleStagePrestartSidecar
+
+	// lifecycleStageMain are tasks without a lifecycle or a lifecycle with an
+	// empty hook value.
+	lifecycleStageMain
+
+	// lifecycleStagePoststartEphemeral are tasks with the "poststart" hook and
+	// sidecar set to "false"
+	lifecycleStagePoststartEphemeral
+
+	// lifecycleStagePoststartSidecar are tasks with the "poststart" hook and
+	// sidecar set to "true".
+	lifecycleStagePoststartSidecar
+
+	// lifecycleStagePoststop are tasks with the "poststop" hook.
+	lifecycleStagePoststop
+)
+
+// Coordinator controls when tasks with a given lifecycle configuration are
+// allowed to start and run.
+//
+// It behaves like a finite state machine where each state transition blocks or
+// allows some task lifecycle types to run.
+type Coordinator struct {
+	logger hclog.Logger
+
+	// tasksByLifecycle is an index used to group and quickly access tasks by
+	// their lifecycle stage.
+	tasksByLifecycle map[lifecycleStage][]string
+
+	// currentState is the current state of the FSM. It must only be accessed
+	// while holding the lock.
+	currentState     coordinatorState
+	currentStateLock sync.RWMutex
+
+	// gates store the gates that control each task lifecycle stage.
+	gates map[lifecycleStage]*Gate
+}
+
+// NewCoordinator returns a new Coordinator with all tasks initially blocked.
+func NewCoordinator(logger hclog.Logger, tasks []*structs.Task, shutdownCh <-chan struct{}) *Coordinator {
+	c := &Coordinator{
+		logger:           logger.Named("task_coordinator"),
+		tasksByLifecycle: indexTasksByLifecycle(tasks),
+		gates:            make(map[lifecycleStage]*Gate),
+	}
+
+	for lifecycle := range c.tasksByLifecycle {
+		c.gates[lifecycle] = NewGate(shutdownCh)
+	}
+
+	c.enterStateLocked(coordinatorStateInit)
+	return c
+}
+
+// Restore is used to set the Coordinator FSM to the correct state when an
+// alloc is restored. Must be called before the allocrunner is running.
+func (c *Coordinator) Restore(states map[string]*structs.TaskState) {
+	// Skip the "init" state when restoring since the tasks were likely already
+	// running, causing the Coordinator to be stuck waiting for them to be
+	// "pending".
+	c.currentStateLock.Lock()
```

Comment:
Not a big deal, but if the AR must not be running, is there any reason to lock?

I ask because if there *is* a reason to lock, then this locking strategy may be unsafe as mutations to the coordinator state and task states may interleave like:

1. Restore Goroutine: enterStateLocked(Prestart)
2. Other Goroutine: enterStateLocked(`<other>`)
3. Other Goroutine: TaskStateUpdated(`<other>`)
4. Restore Goroutine: TaskStateUpdated(states)

So you would end up with incoherent results: the restored task states but some `<other>` coordinator state.

All that to say I'm pretty sure there is no other goroutine manipulating task lifecycles concurrently with Restore, so this isn't a bug, just slightly confusing code.

Reply:
Ah good point. Yeah, there should not be anything else running here, as mentioned in the method godoc so I will remove the lock. Another option would be to have to a `taskStateUpdatedLocked` method that implements the logic of `TaskStateUpdated`, but that seems unnecessary at this point.